channels/rdpgfx/client/rdpgfx_main.c

```diff
@@ -668,14 +598,299 @@ static UINT rdpgfx_recv_evict_cache_entry_pdu(RDPGFX_CHANNEL_CALLBACK* callback,
 	return error;
 }
 
+int rdpgfx_load_cache_import_offer(RDPGFX_PLUGIN* gfx, RDPGFX_CACHE_IMPORT_OFFER_PDU* offer)
+{
+	int index;
+	int count;
+	PERSISTENT_CACHE_ENTRY entry;
+	rdpPersistentCache* persistent = NULL;
+	rdpSettings* settings = gfx->settings;
+
+	offer->cacheEntriesCount = 0;
+
+	if (!settings->BitmapCachePersistEnabled)
+		return 0;
+
+	if (!settings->BitmapCachePersistFile)
+		return 0;
+
+	persistent = persistent_cache_new();
+
+	if (!persistent)
+		return -1;
+
+	if (persistent_cache_open(persistent, settings->BitmapCachePersistFile, FALSE, 3) < 1)
+		goto error;
+
+	if (persistent_cache_get_version(persistent) != 3)
+		goto error;
+
+	count = persistent_cache_get_count(persistent);
+
+	if (count < 1)
+		goto error;
+
+	if (count >= RDPGFX_CACHE_ENTRY_MAX_COUNT)
+		count = RDPGFX_CACHE_ENTRY_MAX_COUNT - 1;
+
+	if (count > gfx->MaxCacheSlots)
+		count = gfx->MaxCacheSlots;
+
+	offer->cacheEntriesCount = (UINT16)count;
+
+	for (index = 0; index < count; index++)
+	{
+		persistent_cache_read_entry(persistent, &entry);
+		offer->cacheEntries[index].cacheKey = entry.key64;
+		offer->cacheEntries[index].bitmapLength = entry.size;
+	}
+
+	persistent_cache_free(persistent);
+
+	return 1;
+error:
+	persistent_cache_free(persistent);
+	return -1;
+}
+
+static int rdpgfx_save_persistent_cache(RDPGFX_PLUGIN* gfx)
+{
+	int index;
+	UINT16 cacheSlot;
+	PERSISTENT_CACHE_ENTRY cacheEntry;
+	rdpPersistentCache* persistent = NULL;
+	rdpSettings* settings = gfx->settings;
+	RdpgfxClientContext* context = (RdpgfxClientContext*)gfx->iface.pInterface;
+
+	if (!context || !context->ExportCacheEntry)
```

Comment:
I'm wondering if it should be `WINPR_ASSERT(context);` ?

Reply:
do we keep asserts in release builds? I see we are using asserts a lot more now, which is fine, but I'd rather keep the null checks in place to avoid null pointer exceptions, even if unlikely to happen in practice.

Reply:
`WINPR_ASSERT`  is currently also defined for (normal) release builds, if you don´t want them (performance critical systems maybe?) you need to define `WITH_VERBOSE_WINPR_ASSERT=OFF`

Reply:
@akallabeth I am not really concerned about the performance cost of asserts in release builds, the issue is that they abort the entire process, which is really a nasty way to report an error condition in production. I understand the asserts are supposed to check for conditions that really shouldn't happen, but sometimes they still do happen anyway.

Reply:
It's not like you can guarantee that stuff that must not happen cannot happen, especially if we're talking about cases where someone would be deliberately sending malicious payloads in hope of triggering asserts in production, resulting in a denial-of-service attack. As for the "you're in trouble anyway", yes, but I'd rather have FreeRDP gracefully terminate the faulty connection instead of terminating the entire process, which in many cases is a lot more than a single-connection RDP client. Think about connection managers hosting multiple RDP client connections at the same time, proxy servers, and RDP servers. Aborting the entire process instead of bubbling up an improbable error condition followed by graceful termination of the faulty connection is not a good idea.

I don't want to this fight this issue, as I said, I'm not so much concerned about the performance cost of the asserts. However, if it's possible to disable the asserts, then I still want those null pointer checks to avoid null pointer exceptions. I don't mind adding asserts, but I want the code to still be properly checking null pointers if the asserts are disabled.

Reply:
@awakecoding the `asserts` are meant for internal stuff (ensure function arguments are valid), of course you should not use that for checks regarding external controlled entities/data.